### ISSUE TITLE
failed to legalize unresolved materialization from ('tensor<1x1x1x16xf32>') to ('tensor<1x64x1x16xf32>') that remained live after conversion

### ISSUE THREAD

**Issue by aijyo** [onnx/onnx-mlir#3321]
# onnx-mlir fails to legalize unresolved materialization when converting DeepFilterNet `erb_dec.onnx`

I’m using **onnx-mlir** to convert DeepFilterNet’s `erb_dec.onnx` into a library and encountered the following error:

```
[2/3] Wed Oct 29 13:22:23 2025 (0s) Compiling and Optimizing MLIR Module
failed to legalize unresolved materialization from ('tensor<1x1x1x16xf32>') to ('tensor<1x64x1x16xf32>') that remained live after conversion
```

From the DeepFilterNet source, I suspect the issue is caused by an implicit broadcast in this part of the model:

```python
def forward(self, emb: Tensor, e3: Tensor, e2: Tensor, e1: Tensor, e0: Tensor) -> Tensor:
    # Estimates erb mask
    b, _, t, f8 = e3.shape
    emb, _ = self.emb_gru(emb)
    emb = emb.view(b, t, f8, -1).permute(0, 3, 1, 2)  # [B, C*8, T, F/8]
    e3 = self.convt3(self.conv3p(e3) + emb)          # [B, C*4, T, F/4]
    e2 = self.convt2(self.conv2p(e2) + e3)           # [B, C*2, T, F/2]
    e1 = self.convt1(self.conv1p(e1) + e2)           # [B, C, T, F]
    m  = self.conv0_out(self.conv0p(e0) + e1)        # [B, 1, T, F]
    return m
```

**Constraint:** I currently cannot modify the source to re-export an ONNX without implicit broadcasts.

**Question:** Is there any way to handle this class of issue with **onnx-mlir** (or by post-processing the existing ONNX) so that the materialization from `tensor<1x1x1x16xf32>` to `tensor<1x64x1x16xf32>` is legalized? Any recommended passes, flags, or graph edits to make this work **without changing the original model code** would be greatly appreciated. Thanks!

**Comment by chentong319** [comment]
Which ONNX op created this cast? I am curious about the source.
If the implicit broadcast can be detected within ONNX op, you may add a transformation to fix it with onnx.expand operation.

**Comment by aijyo** [comment]
[dfn_onnx_rewriter.py](https://github.com/user-attachments/files/23248477/dfn_onnx_rewriter.py)

Thank you. I solved the issue using the method you suggested, but another problem might have been introduced or triggered by other causes after resolving this one. As a beginner, is it possible to join the Slack group directly? I feel there is a lot to learn. This script I used for conversion is on Python (it processes deepfilternet3_ll 0.5.6 version of erb_dec.onnx). The command is:
python dfn_onnx_rewriter.py --in erb_dec.onnx --out erb_dec_fixed.onnx --opset 17 --freeze-dim S=1 --freeze-dim B=1 --freeze-dim T=1
If you have time, could you take a look? After generating erb_dec_fixed.onnx with this script, I then run:
onnx-mlir.exe --EmitONNXIR ./erb_dec_fixed.onnx -o erb_dec_fixed
and it triggers the following error. The log locations I printed come from my modifications to the ONNX-MLIR source code, and I currently cannot trace which step causes the problem. Many thanks again in advance!

[1/3] Fri Oct 31 10:22:14 2025 (0s) Importing ONNX Model to MLIR Module from "erb_dec_fixed.onnx"
[2/3] Fri Oct 31 10:22:14 2025 (0s) Compiling and Optimizing MLIR Module
[ONNXConv verify fail] node=<none> loc=loc("/convt2/0/ConvTranspose") M=1 Cpg=64 g=64 Cin=64
'onnx.Conv' op Channel Out (M) must be a multiple of the number of groups
'onnx.Conv' op verification failed
[ONNXConv verify fail] node=<none> loc=loc("/convt2/0/ConvTranspose") M=1 Cpg=64 g=64 Cin=64
'onnx.Conv' op Channel Out (M) must be a multiple of the number of groups
'onnx.Conv' op verification failed
'onnx.Conv' op Channel In (C) of input must be equal 2nd dim of weights times g
'onnx.Conv' op verification failed

**Comment by aijyo** [comment]
> Which ONNX op created this cast? I am curious about the source. If the implicit broadcast can be detected within ONNX op, you may add a transformation to fix it with onnx.expand operation.



[erb_dec_fixed.onnx](https://github.com/user-attachments/files/23251093/erb_dec_fixed.onnx.txt)


I have discovered a very strange issue, which seems to be caused by a bug in the program logic. I used the command
--EmitLib ./erb_dec_fixed.onnx -o erb_dec_fixed --mlir-print-ir-after-all --mlir-elide-elementsattrs-if-larger=16 --mlir-print-op-on-diagnostic
to generate [erb_dec_emit_lib_output.txt](https://github.com/user-attachments/files/23251081/erb_dec_emit_lib_output.txt). In that file, I saw
%58 = onnx.Constant dense<__elided__> : tensor<64x1x1x3xf32>
and
%507 = "onnx.ConvTranspose"(%506, %58, %46) {auto_pad = "NOTSET", dilations = [1, 1], group = 64 : si64, kernel_shape = [1, 3], onnx_node_name = "/convt2/0/ConvTranspose"} This means that %58 is used as the W input for /convt2/0/ConvTranspose, and its shape is 64x1x1x3xf32, i.e., M=64 (shape[0]).

However, during debugging, I found that the output was: [ONNXConv verify fail] node=<none> loc=loc("/convt2/0/ConvTranspose") ConvTranspose w0=1 ConvTranspose w1=64 ConvTranspose w2=1 ConvTranspose w3=3 (this is a log I added myself when debugging onnx-mlir; I will include my locally modified LogicalResult ONNXConvOp::verify() at the end).

In other words, at this point, the shape of W is no longer the original 64x1x1x3xf32, but instead has become 1x64x1x3xf32. This triggers the exception 'onnx.Conv' op Channel Out (M) must be a multiple of the number of groups.
Can you help me investigate the cause of this issue? I’m really at my wits’ end and have to turn to you for help. Thank you.
Below is the code I modified to add debug logs:

C++

LogicalResult ONNXConvOp::verify() {
  ONNXConvOpAdaptor operandAdaptor = ONNXConvOpAdaptor(*this);
  // Get operands.
  auto X = operandAdaptor.getX();
  auto W = operandAdaptor.getW();
  auto B = operandAdaptor.getB();
  bool hasBias = !isNoneValue(B);
  int64_t g = getGroup();
  if (g < 1)
    return emitOpError("group must be strictly positive");
  // Get spatial rank.
  if (!hasShapeAndRank(W)) {
    // Won't be able to do any checking at this stage.
    return success();
  }
  auto wShape = mlir::cast<ShapedType>(W.getType()).getShape();
  int64_t spatialRank = wShape.size() - 2;
  // If ranked, verify ranks of inputs.
  if (spatialRank < 1)
    return emitOpError("Spatial rank must be strictly positive");

  if (wShape[0] != ShapedType::kDynamic && wShape[0] % g != 0) {
    // This rule is not enforced in the spec but is present in Keras,
    // Pytorch, and simplifies the code.
    // Note: Pytorch requires both channel in (CI) and channel out (CO) to be
    // multiple of group number (G).
    // https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html
    // ONNX clearly states that C (channel in or CI here) is a multiple of group
    // number (G).
    // https://github.com/onnx/onnx/blob/main/docs/Operators.md#Conv
    // Quote: X.shape[1] == (W.shape[1] * group) == C
    // Keras also specifies it: Input channels and filters must both be
    // divisible by groups.
    // https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D
    // --- DEBUG: snapshot & print failing Conv (safe for MSVC) ---
    auto *op = (*this).getOperation();

    auto w0 = wShape[0];
    auto w1 = wShape[1];
    auto w2 = wShape[2];
    auto w3 = wShape[3];
    // Build a readable location string (no dyn_cast needed)
    std::string locStr;
    {
      llvm::raw_string_ostream os(locStr);
      op->getLoc().print(
          os); // e.g., "builtin.name_loc(@/conv0_out/1/Conv:...)" or
               // fused/file:line:col
    }

    // Extract key dims without member dyn_cast (use llvm::dyn_cast on Type)
    long long g = this->getGroup();

    long long M = -1, Cpg = -1;
    if (auto wTy =
            llvm::dyn_cast<mlir::RankedTensorType>(this->getW().getType())) {
      if (wTy.hasStaticShape() && wTy.getRank() == 4) {
        auto shp = wTy.getShape();
        M = shp[0];   // Channel Out
        Cpg = shp[1]; // C_per_group
      }
    }

    long long Cin = -1;
    if (auto xTy =
            llvm::dyn_cast<mlir::RankedTensorType>(this->getX().getType())) {
      if (xTy.getRank() > 1) {
        auto d1 = xTy.getDimSize(1); // negative if dynamic
        if (d1 >= 0)
          Cin = d1; // Channel In
      }
    }

    // Optional: original name attribute (might be absent)
    mlir::StringAttr nameAttr =
        op->getAttrOfType<mlir::StringAttr>("onnx_node_name");

    // One-line debug print
    llvm::errs() << "[ONNXConv verify fail] "
                 << "node="
                 << (nameAttr ? nameAttr.getValue() : mlir::StringRef("<none>"))
                 << " loc=" << locStr << " ConvTranspose w0=" << w0 << " ConvTranspose w1=" << w1
                 << " ConvTranspose w2=" << w2 << " ConvTranspose w3=" << w3 << "\n";

    llvm::errs() << "[ONNXConv verify fail] "
                 << "node="
                 << (nameAttr ? nameAttr.getValue() : mlir::StringRef("<none>"))
                 << " loc=" << locStr << " M=" << M << " Cpg=" << Cpg
                 << " g=" << g << " Cin=" << Cin << "\n";
    // --- END DEBUG ---



    return emitOpError(
        "Channel Out (M) must be a multiple of the number of groups");
  }
  if (hasShapeAndRank(X)) {
    auto xShape = mlir::cast<ShapedType>(X.getType()).getShape();
    if (static_cast<int64_t>(xShape.size()) - 2 != spatialRank)
      return emitOpError("Input and filter rank mismatch");
    if (xShape[1] != ShapedType::kDynamic && xShape[1] % g != 0)
      return emitOpError(
          "Channel In (C) must be a multiple of the number of groups");
    if (xShape[1] != ShapedType::kDynamic &&
        wShape[1] != ShapedType::kDynamic && xShape[1] != wShape[1] * g) {
      return emitOpError("Channel In (C) of input must be equal 2nd dim "
                         "of weights times g");
    }
  }
  if (hasBias && hasShapeAndRank(B)) {
    auto bShape = mlir::cast<ShapedType>(B.getType()).getShape();
    if (bShape.size() != 1)
      return emitOpError("Bias should have a rank of one");
    if (bShape[0] != ShapedType::kDynamic &&
        wShape[0] != ShapedType::kDynamic && wShape[0] != bShape[0])
      return emitOpError(
          "Bias should have same dimension as first dimension of weights");
  }
  // Verify parameters.
  if (failed(verifyKernelShape<ONNXConvOp>(
          this, W, getKernelShape(), spatialRank)))
    return failure();
  if (failed(verifyStrides<ONNXConvOp>(this, spatialRank)))
    return failure();
  if (failed(verifyDilations<ONNXConvOp>(this, spatialRank)))
    return failure();
  if (failed(verifyPadding<ONNXConvOp>(this, spatialRank)))
    return failure();
  return success();
}
